File: src/pipeline/shift-event-detector.js
// shift-event-detector — pipeline streaming que detecta eventos de troca de
// marcha e emite o evento completo (25 campos da spec).
//
// Uso:
//   const det = createShiftEventDetector({
//     piloto_id, carro_id, sessao_id, car,
//     onEvent: (ev) => saveEvent(ev),
//     resolveTrecho: (sample) => 'trecho-x' | null
//   });
//   det.pushSample(sample);
//
// O detector mantém um buffer de samples recentes. Quando vê uma queda/subida
// brusca de RPM com velocidade aproximadamente constante, marca o instante e,
// após `windowAfter` ms, emite o evento construído com os 3 pontos
// (rpm_before médio, rpm_at_shift, rpm_after médio) + estimativa de marcha
// + alvo + delta + flags.

import { estimateGear } from '../domain/gear-estimation.js';
import { computeShiftTarget } from '../domain/shift-target.js';

const DEFAULTS = Object.freeze({
  windowBefore: 200,     // ms — janela pra média rpm_before / accel_before
  windowAfter: 200,      // ms — janela pra média rpm_after / accel_after
  rpmDrop: 800,          // queda mínima pra ser upshift
  rpmRise: 600,          // subida mínima pra ser downshift
  speedTolPct: 0.05,     // velocidade não pode variar mais que 5% durante shift
  cooldownMs: 400,       // ms — bloqueia detecção dupla logo em seguida
  maxDtMs: 250,          // janela máxima entre samples consecutivos pra contar como shift
  tpsMin: 10             // ignora samples com TPS < 10 (provável neutro/embreagem)
});

export function createShiftEventDetector(opts = {}) {
  const cfg = { ...DEFAULTS, ...opts };
  const {
    piloto_id = null, carro_id = null, sessao_id = null,
    car = null,
    onEvent = () => {},
    resolveTrecho = null
  } = cfg;

  let state = 'idle';     // 'idle' | 'pending' | 'cooldown'
  let buffer = [];
  let pending = null;
  let cooldownUntil = 0;
  const seenKeys = new Set();
  const emitted = [];

  function pushSample(sample) {
    if (!sample
      || !Number.isFinite(sample.timestamp)
      || !Number.isFinite(sample.rpm)
      || !Number.isFinite(sample.speed)) return;

    buffer.push(sample);
    const keepSince = sample.timestamp - (cfg.windowBefore + cfg.windowAfter + 500);
    while (buffer.length && buffer[0].timestamp < keepSince) buffer.shift();

    if (state === 'cooldown') {
      if (sample.timestamp >= cooldownUntil) state = 'idle';
      else return;
    }

    if (state === 'idle') {
      if (buffer.length < 2) return;
      const prev = buffer[buffer.length - 2];
      const cur = sample;
      const dt = cur.timestamp - prev.timestamp;
      if (dt <= 0 || dt > cfg.maxDtMs) return;
      const dRpm = cur.rpm - prev.rpm;
      const speedRel = Math.abs(cur.speed - prev.speed) / Math.max(prev.speed, 1);
      if (speedRel > cfg.speedTolPct) return;
      let type = null;
      if (dRpm <= -cfg.rpmDrop) type = 'up';
      else if (dRpm >= cfg.rpmRise) type = 'down';
      if (!type) return;
      pending = { type, atTimestamp: cur.timestamp, beforeSample: prev, atSample: cur };
      state = 'pending';
      return;
    }

    if (state === 'pending') {
      const elapsed = sample.timestamp - pending.atTimestamp;
      if (elapsed < cfg.windowAfter) return;
      const event = buildEvent({
        cfg, buffer, pending, finalSample: sample,
        piloto_id, carro_id, sessao_id, car, resolveTrecho
      });
      pending = null;
      cooldownUntil = sample.timestamp + cfg.cooldownMs;
      state = 'cooldown';
      if (event) {
        if (!seenKeys.has(event.dedup_key)) {
          seenKeys.add(event.dedup_key);
          emitted.push(event);
          try { onEvent(event); } catch { /* callback não derruba pipeline */ }
        }
      }
    }
  }

  return {
    pushSample,
    _state: () => state,
    _emitted: () => emitted.slice()
  };
}

function buildEvent({ cfg, buffer, pending, finalSample, piloto_id, carro_id, sessao_id, car, resolveTrecho }) {
  const before = pending.beforeSample;
  const at = pending.atSample;
  const after = finalSample;

  const tps = Number.isFinite(at.tps) ? at.tps : null;
  if (tps != null && tps < cfg.tpsMin) return null;

  const tCutBefore = at.timestamp - cfg.windowBefore;
  const beforeWindow = buffer.filter(s => s.timestamp >= tCutBefore && s.timestamp < at.timestamp);
  const afterWindow = buffer.filter(s => s.timestamp > at.timestamp && s.timestamp <= after.timestamp);

  const avg = (arr, pick) => {
    const vals = arr.map(pick).filter(v => Number.isFinite(v));
    return vals.length ? vals.reduce((a, b) => a + b, 0) / vals.length : null;
  };

  const rpm_before_avg = avg(beforeWindow, s => s.rpm);
  const rpm_after_avg = avg(afterWindow, s => s.rpm);
  // No upshift, rpm_at_shift é o rpm imediatamente antes da queda.
  // No downshift, é imediatamente após a subida (a estamparia da nova marcha).
  const rpm_at_shift = pending.type === 'up' ? before.rpm : at.rpm;

  const calibration = car && car.gear_signatures
    ? { gear_signatures: car.gear_signatures }
    : null;
  const gearBeforeRes = calibration
    ? estimateGear(before, beforeWindow, calibration)
    : { gear: null, confidence: 0 };
  const gearAfterRes = calibration
    ? estimateGear(after, afterWindow, calibration)
    : { gear: null, confidence: 0 };

  let target = null;
  if (car && Number.isFinite(car.redline_rpm)) {
    target = computeShiftTarget({
      car,
      gear: gearBeforeRes.gear,
      gearConfidence: gearBeforeRes.confidence
    });
  }
  const target_optimal_rpm = target ? target.optimalRpm : null;
  const target_visual_rpm = target_optimal_rpm; // Bloco 5 ajusta com reaction
  const delta_rpm = (Number.isFinite(rpm_at_shift) && Number.isFinite(target_optimal_rpm))
    ? rpm_at_shift - target_optimal_rpm
    : null;

  const tolerance = (car && Number.isFinite(car.tolerance_rpm)) ? car.tolerance_rpm : 150;
  let status = 'unknown';
  if (Number.isFinite(delta_rpm)) {
    status = Math.abs(delta_rpm) <= tolerance ? 'green' : 'red';
  }

  let mode_used = 'SAFE_MODE';
  if (Array.isArray(car?.dyno_curve) && car.dyno_curve.length > 0) mode_used = 'DYNO_CALIBRATED';
  else if (target && target.source === 'learned') mode_used = 'TELEMETRY_LEARNED';

  const overrev_flag = (Number.isFinite(rpm_at_shift) && car && Number.isFinite(car.redline_rpm))
    ? rpm_at_shift > car.redline_rpm
    : false;

  let trecho_id = null;
  if (typeof resolveTrecho === 'function') {
    try {
      const v = resolveTrecho(at);
      if (typeof v === 'string' && v.length > 0) trecho_id = v;
    } catch {
      trecho_id = null;
    }
  }

  const data_inconsistent_flag = (
    rpm_before_avg == null
    || rpm_after_avg == null
    || !Number.isFinite(rpm_at_shift)
    || tps == null
  );

  const dedup_key = [
    sessao_id || 'no-session',
    at.timestamp,
    gearBeforeRes.gear ?? 'x',
    gearAfterRes.gear ?? 'x'
  ].join(':');

  return {
    piloto_id,
    carro_id,
    sessao_id,
    volta_id: null,
    trecho_id,
    timestamp: at.timestamp,
    rpm_before: rpm_before_avg != null ? Math.round(rpm_before_avg) : null,
    rpm_at_shift: Number.isFinite(rpm_at_shift) ? Math.round(rpm_at_shift) : null,
    rpm_after: rpm_after_avg != null ? Math.round(rpm_after_avg) : null,
    speed_kmh: Number.isFinite(at.speed) ? at.speed : null,
    tps,
    gear_before: gearBeforeRes.gear,
    gear_after: gearAfterRes.gear,
    gear_confidence: gearAfterRes.confidence,
    accel_long_before: avg(beforeWindow, s => s?.accel?.x),
    accel_long_during: Number.isFinite(at?.accel?.x) ? at.accel.x : null,
    accel_long_after: avg(afterWindow, s => s?.accel?.x),
    target_optimal_rpm,
    target_visual_rpm,
    delta_rpm,
    status,
    mode_used,
    overrev_flag,
    data_inconsistent_flag,
    lesson_text: null,
    dedup_key
  };
}

File: ios/p1fast-core/Sources/P1FastCore/ShiftEventDetector.swift
// ═══════════════════════════════════════════════════════════
// ShiftEventDetector — port Swift de:
//   src/pipeline/shift-event-detector.js
// ═══════════════════════════════════════════════════════════
// Pipeline streaming que detecta eventos de troca de marcha
// (upshift/downshift) e emite ShiftEvent completo.
//
// Estado: idle → pending → cooldown → idle (ciclo).
// Em pending acumula janela `windowAfter` antes de emitir.
//
// MS-8.5 (port 2026-05-04). Fecha o pipeline shift-light em Swift
// junto com `ShiftLightBridge`. Em Tier 0 (sem RPM real) o bridge
// descarta samples sem RPM antes de chegarem aqui.

import Foundation

public enum ShiftDetectorConst {
    public static let windowBefore: Double  = 200   // ms
    public static let windowAfter: Double   = 200
    public static let rpmDrop: Double       = 800   // upshift mínimo
    public static let rpmRise: Double       = 600   // downshift mínimo
    public static let speedTolPct: Double   = 0.05  // 5%
    public static let cooldownMs: Double    = 400
    public static let maxDtMs: Double       = 250
    public static let tpsMin: Double        = 10
}

/// Sample alimentado pelo bridge — contrato do detector.
public struct ShiftDetectorSample: Sendable, Equatable {
    public var rpm: Double
    public var speed: Double         // km/h
    public var tps: Double?
    public var accelX: Double        // long
    public var accelY: Double
    public var accelZ: Double
    public var timestamp: Double     // ms

    public init(
        rpm: Double, speed: Double,
        tps: Double? = nil,
        accelX: Double = 0, accelY: Double = 0, accelZ: Double = 0,
        timestamp: Double
    ) {
        self.rpm = rpm
        self.speed = speed
        self.tps = tps
        self.accelX = accelX
        self.accelY = accelY
        self.accelZ = accelZ
        self.timestamp = timestamp
    }
}

public enum ShiftType: String, Sendable, Codable {
    case up
    case down
}

public enum ShiftMode: String, Sendable, Codable {
    case dynoCalibrated  = "DYNO_CALIBRATED"
    case telemetryLearned = "TELEMETRY_LEARNED"
    case safeMode        = "SAFE_MODE"
}

public struct ShiftEventEmitted: Sendable, Equatable {
    public var pilotoId: String?
    public var carroId: String?
    public var sessaoId: String?
    public var voltaId: String?
    public var trechoId: String?
    public var timestamp: Double
    public var rpmBefore: Double?
    public var rpmAtShift: Double?
    public var rpmAfter: Double?
    public var speedKmh: Double?
    public var tps: Double?
    public var gearBefore: Int?
    public var gearAfter: Int?
    public var gearConfidence: Double
    public var accelLongBefore: Double?
    public var accelLongDuring: Double?
    public var accelLongAfter: Double?
    public var targetOptimalRpm: Double?
    public var targetVisualRpm: Double?
    public var deltaRpm: Double?
    public var status: String        // "green" | "red" | "unknown"
    public var modeUsed: ShiftMode
    public var overrevFlag: Bool
    public var dataInconsistentFlag: Bool
    public var lessonText: String?
    public var dedupKey: String
}

public final class ShiftEventDetector {
    // Configuração
    public var pilotoId: String?
    public var carroId: String?
    public var sessaoId: String?
    public var car: ShiftCar?
    public var resolveTrecho: ((ShiftDetectorSample) -> String?)?
    public var onEvent: ((ShiftEventEmitted) -> Void)?

    // Estado interno
    private enum State { case idle, pending, cooldown }
    private var state: State = .idle
    private var buffer: [ShiftDetectorSample] = []
    private var pendingType: ShiftType?
    private var pendingAt: Double = 0
    private var pendingBefore: ShiftDetectorSample?
    private var pendingAtSample: ShiftDetectorSample?
    private var cooldownUntil: Double = 0
    private var seenKeys: Set<String> = []
    private var emittedEvents: [ShiftEventEmitted] = []

    public init(
        pilotoId: String? = nil,
        carroId: String? = nil,
        sessaoId: String? = nil,
        car: ShiftCar? = nil,
        onEvent: ((ShiftEventEmitted) -> Void)? = nil,
        resolveTrecho: ((ShiftDetectorSample) -> String?)? = nil
    ) {
        self.pilotoId = pilotoId
        self.carroId = carroId
        self.sessaoId = sessaoId
        self.car = car
        self.onEvent = onEvent
        self.resolveTrecho = resolveTrecho
    }

    public func pushSample(_ sample: ShiftDetectorSample) {
        guard sample.rpm.isFinite, sample.speed.isFinite, sample.timestamp.isFinite else { return }

        buffer.append(sample)
        let keepSince = sample.timestamp - (ShiftDetectorConst.windowBefore + ShiftDetectorConst.windowAfter + 500)
        while let first = buffer.first, first.timestamp < keepSince {
            buffer.removeFirst()
        }

        if state == .cooldown {
            if sample.timestamp >= cooldownUntil { state = .idle }
            else { return }
        }

        if state == .idle {
            guard buffer.count >= 2 else { return }
            let prev = buffer[buffer.count - 2]
            let cur = sample
            let dt = cur.timestamp - prev.timestamp
            guard dt > 0, dt <= ShiftDetectorConst.maxDtMs else { return }
            let dRpm = cur.rpm - prev.rpm
            let speedRel = abs(cur.speed - prev.speed) / max(prev.speed, 1)
            guard speedRel <= ShiftDetectorConst.speedTolPct else { return }
            let type: ShiftType?
            if dRpm <= -ShiftDetectorConst.rpmDrop      { type = .up }
            else if dRpm >= ShiftDetectorConst.rpmRise  { type = .down }
            else                                        { type = nil }
            guard let t = type else { return }
            pendingType = t
            pendingAt = cur.timestamp
            pendingBefore = prev
            pendingAtSample = cur
            state = .pending
            return
        }

        if state == .pending {
            let elapsed = sample.timestamp - pendingAt
            guard elapsed >= ShiftDetectorConst.windowAfter else { return }
            let event = buildEvent(finalSample: sample)
            pendingType = nil
            pendingBefore = nil
            pendingAtSample = nil
            cooldownUntil = sample.timestamp + ShiftDetectorConst.cooldownMs
            state = .cooldown
            if let ev = event {
                if !seenKeys.contains(ev.dedupKey) {
                    seenKeys.insert(ev.dedupKey)
                    emittedEvents.append(ev)
                    onEvent?(ev)
                }
            }
        }
    }

    public var emitted: [ShiftEventEmitted] { emittedEvents }
    public var debugState: String {
        switch state {
        case .idle: return "idle"
        case .pending: return "pending"
        case .cooldown: return "cooldown"
        }
    }

    // MARK: - Build event

    private func buildEvent(finalSample: ShiftDetectorSample) -> ShiftEventEmitted? {
        guard let type = pendingType,
              let before = pendingBefore,
              let at = pendingAtSample else { return nil }
        let after = finalSample

        let tps = at.tps
        if let t = tps, t < ShiftDetectorConst.tpsMin { return nil }

        let tCutBefore = at.timestamp - ShiftDetectorConst.windowBefore
        let beforeWindow = buffer.filter { $0.timestamp >= tCutBefore && $0.timestamp < at.timestamp }
        let afterWindow  = buffer.filter { $0.timestamp > at.timestamp && $0.timestamp <= after.timestamp }

        let rpmBeforeAvg = avg(beforeWindow.map(\.rpm))
        let rpmAfterAvg  = avg(afterWindow.map(\.rpm))
        // No upshift, rpm_at_shift é o rpm imediatamente antes da queda.
        // No downshift, é imediatamente após a subida.
        let rpmAtShift: Double = (type == .up) ? before.rpm : at.rpm

        // Estimativa de marcha (antes/depois) usando GearEstimation
        var gearBefore: GearEstimate? = nil
        var gearAfter: GearEstimate? = nil
        if let signatures = car?.gearSignatures, !signatures.isEmpty {
            let parsedSignatures: [Int: GearSignature] = signatures.reduce(into: [:]) { acc, kv in
                if let g = Int(kv.key) { acc[g] = kv.value }
            }
            let beforeInputs = beforeWindow.map { toGearInput($0) }
            let afterInputs  = afterWindow.map { toGearInput($0) }
            gearBefore = GearEstimation.estimateGear(
                sample: toGearInput(before),
                history: beforeInputs,
                signatures: parsedSignatures
            )
            gearAfter = GearEstimation.estimateGear(
                sample: toGearInput(after),
                history: afterInputs,
                signatures: parsedSignatures
            )
        }

        // ShiftTarget
        var targetOptimalRpm: Double? = nil
        var targetVisualRpm: Double? = nil
        if let car {
            let result = (try? ShiftTarget.computeShiftTarget(
                car: car,
                gear: gearBefore?.gear ?? 0,
                gearConfidence: gearBefore?.confidence
            ))
            if let r = result {
                targetOptimalRpm = r.optimalRpm
                targetVisualRpm = r.visualRpm
            }
        }
        let deltaRpm: Double? = (rpmAtShift.isFinite && targetOptimalRpm != nil)
            ? rpmAtShift - targetOptimalRpm!
            : nil

        let tolerance = car?.effectiveToleranceRpm ?? CarShiftDefaults.toleranceRpm
        let status: String
        if let d = deltaRpm {
            status = abs(d) <= tolerance ? "green" : "red"
        } else {
            status = "unknown"
        }

        var modeUsed: ShiftMode = .safeMode
        if let car, !(car.dynoCurve?.isEmpty ?? true) { modeUsed = .dynoCalibrated }
        else if (car?.learnedTargets?[gearBefore?.gear ?? -1]) != nil { modeUsed = .telemetryLearned }

        let overrevFlag: Bool = {
            guard let car, car.redlineRpm.isFinite else { return false }
            return rpmAtShift > car.redlineRpm
        }()

        let trechoId: String? = resolveTrecho?(at)

        let dataInconsistentFlag = (
            rpmBeforeAvg == nil
            || rpmAfterAvg == nil
            || !rpmAtShift.isFinite
            || tps == nil
        )

        let dedupKey = [
            sessaoId ?? "no-session",
            "\(at.timestamp)",
            gearBefore?.gear.map { "\($0)" } ?? "x",
            gearAfter?.gear.map { "\($0)" } ?? "x"
        ].joined(separator: ":")

        return ShiftEventEmitted(
            pilotoId: pilotoId,
            carroId: carroId,
            sessaoId: sessaoId,
            voltaId: nil,
            trechoId: trechoId,
            timestamp: at.timestamp,
            rpmBefore: rpmBeforeAvg.map { $0.rounded() },
            rpmAtShift: rpmAtShift.isFinite ? rpmAtShift.rounded() : nil,
            rpmAfter: rpmAfterAvg.map { $0.rounded() },
            speedKmh: at.speed.isFinite ? at.speed : nil,
            tps: tps,
            gearBefore: gearBefore?.gear,
            gearAfter: gearAfter?.gear,
            gearConfidence: gearAfter?.confidence ?? 0,
            accelLongBefore: avg(beforeWindow.map(\.accelX)),
            accelLongDuring: at.accelX.isFinite ? at.accelX : nil,
            accelLongAfter: avg(afterWindow.map(\.accelX)),
            targetOptimalRpm: targetOptimalRpm,
            targetVisualRpm: targetVisualRpm,
            deltaRpm: deltaRpm,
            status: status,
            modeUsed: modeUsed,
            overrevFlag: overrevFlag,
            dataInconsistentFlag: dataInconsistentFlag,
            lessonText: nil,
            dedupKey: dedupKey
        )
    }
}

// MARK: - Bridge (shift-light-bridge.js port)

/// Adapta o stream do MobileTelemetry (Sample canônico) para o
/// input do `ShiftEventDetector`. Sample sem RPM é descartado
/// (princípio "não fabricar dados" — Tier 0 dormente).
public final class ShiftLightBridge {
    public typealias RpmGetter = (ShiftBridgeSample) -> Double?
    public typealias TpsGetter = (ShiftBridgeSample) -> Double?

    private let detector: ShiftEventDetector
    private let getRpm: RpmGetter
    private let getTps: TpsGetter
    private var lastTs: Double = -1

    public init(
        detector: ShiftEventDetector,
        getRpm: @escaping RpmGetter,
        getTps: @escaping TpsGetter = { _ in nil }
    ) {
        self.detector = detector
        self.getRpm = getRpm
        self.getTps = getTps
    }

    public func onTelemetrySample(_ sample: ShiftBridgeSample) {
        let ts = sample.tMono ?? sample.t
        guard ts.isFinite, ts > lastTs else { return }
        lastTs = ts
        guard let rpm = getRpm(sample), rpm.isFinite, rpm > 0 else { return }
        let speed: Double?
        if let kmh = sample.kmh, kmh.isFinite { speed = kmh }
        else if let s = sample.speed, s.isFinite { speed = s * 3.6 }
        else { speed = nil }
        guard let sp = speed else { return }
        let tps = getTps(sample)
        let tpsFinite = (tps?.isFinite == true) ? tps : nil
        detector.pushSample(ShiftDetectorSample(
            rpm: rpm,
            speed: sp,
            tps: tpsFinite,
            accelX: sample.accX ?? 0,
            accelY: sample.accY ?? 0,
            accelZ: sample.accZ ?? 0,
            timestamp: ts
        ))
    }

    public func reset() { lastTs = -1 }
}

/// Sample canônico do bridge — paridade com o Sample JS do MobileTelemetry.
public struct ShiftBridgeSample: Sendable, Equatable {
    public var t: Double
    public var tMono: Double?
    public var kmh: Double?
    public var speed: Double?       // m/s
    public var accX: Double?
    public var accY: Double?
    public var accZ: Double?

    public init(
        t: Double, tMono: Double? = nil,
        kmh: Double? = nil, speed: Double? = nil,
        accX: Double? = nil, accY: Double? = nil, accZ: Double? = nil
    ) {
        self.t = t
        self.tMono = tMono
        self.kmh = kmh
        self.speed = speed
        self.accX = accX
        self.accY = accY
        self.accZ = accZ
    }
}

// MARK: - Helpers

private func avg(_ vals: [Double]) -> Double? {
    let valid = vals.filter { $0.isFinite }
    guard !valid.isEmpty else { return nil }
    return valid.reduce(0, +) / Double(valid.count)
}

private func toGearInput(_ s: ShiftDetectorSample) -> GearSampleInput {
    GearSampleInput(
        rpm: s.rpm, speed: s.speed,
        tps: s.tps, accelLong: s.accelX,
        timestamp: s.timestamp
    )
}
